### ISSUE TITLE
About unity version

### ISSUE THREAD

**Issue by Winvz** [RenderHeads/UnityPlugin-AVProVideo#1687]
I saw that the Unity store page requires Unity2019 and above, but the introduction is supporting Unity2018 and above, which shall prevail?

Also, what is the difference between the trial version and the official version

**Comment by Winvz** [comment]
Does the trial version support Android and ios platforms

**Comment by Chris-RH** [comment]
Under Unity guidelines (https://assetstore.unity.com/publishing--release-updates), the minimum Unity version that we can use to publish assets in the asset store is 2019.4 (2021.3 for new assets). 
You can absolutely use Unity 2018 in your project, but the latest version is still well over 2 years old.  Is there a specific feature of Unity 2018 that you need for your project that is no longer available in later version of Unity?

The Trial version is the same as the Ultra version, apart from it doesn't support OES textures and it is watermarked.  The trial version supports Android, iOS, macOS, tvOS and Windows.  More info here: https://renderheads.com/content/docs/AVProVideo/articles/download.html#editions

**Comment by Winvz** [comment]
I used Unity2018 to create an empty project, then imported the trial version of unitypackage, used Mediaplay as the startup scene, directly generated Apk, and found an error when running on the simulator
2023-11-14 17:15:20.722 24772-24793/? E/Unity: AndroidJavaException: java.lang.IllegalArgumentException: Can't allocate an array of void
    java.lang.IllegalArgumentException: Can't allocate an array of void
        at java.lang.reflect.Array.newArray(Array.java:794)
        at java.lang.reflect.Array.newInstance(Array.java:78)
        at com.unity3d.player.ReflectionHelper.a(Unknown Source:136)
        at com.unity3d.player.ReflectionHelper.a(Unknown Source:19)
        at com.unity3d.player.ReflectionHelper.getMethodID(Unknown Source:18)
        at com.unity3d.player.UnityPlayer.nativeRender(Native Method)
        at com.unity3d.player.UnityPlayer.e(Unknown Source:0)
        at com.unity3d.player.UnityPlayer$f$1.handleMessage(Unknown Source:148)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:193)
        at com.unity3d.player.UnityPlayer$f.run(Unknown Source:20)
      at UnityEngine._AndroidJNIHelper.GetMethodID (System.IntPtr jclass, System.String methodName, System.String signature, System.Boolean isStatic) [0x00041] in <e4e4a59cfd2e4de5b5e42405672f4bca>:0 
      at UnityEngi

**Comment by Winvz** [comment]
Thank you for your reply.
1.The same results would occur for the Samsung A70s
2.Thunder Simulator, Android 9.0
3.This is a 5 year old project and upgrading Unity is not an easy task, but we did consider upgrading to Unity2019 recently
My intuition tells me that this error only appears on Unity 2018 exported APKs

**Comment by Winvz** [comment]
4. Can I cache and play the same video on Android/iOS?
An error occurred when I used the Apk generated by Unity2019 to do this:
> 2023-11-16 15:42:30.797 2238-2300/? E/Unity: OPENGL NATIVE PLUG-IN ERROR: GL_INVALID_FRAMEBUFFER_OPERATION:  Framebuffer is not complete or incompatible with command
(Filename: ./Runtime/GfxDevice/opengles/GfxDeviceGLES.cpp Line: 442)

**Comment by Winvz** [comment]
The fourth problem seems to be that of the exoplayer
> 2023-11-16 16:20:06.770 14485-14540/? E/ExoPlayerImplInternal: Playback error
      com.google.android.exoplr2avp.ExoPlaybackException: Source error
        at com.google.android.exoplr2avp.ExoPlayerImplInternal.handleIoException(ExoPlayerImplInternal.java:684)
        at com.google.android.exoplr2avp.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:656)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
     Caused by: com.google.android.exoplr2avp.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 416
        at com.google.android.exoplr2avp.upstream.DefaultHttpDataSource.open(DefaultHttpDataSource.java:419)
        at com.google.android.exoplr2avp.upstream.TeeDataSource.open(TeeDataSource.java:60)
        at com.google.android.exoplr2avp.upstream.cache.CacheDataSource.openNextSource(CacheDataSource.java:802)
        at com.google.android.exoplr2avp.upstream.cache.CacheDataSource.read(CacheDataSource.java:657)
        at com.google.android.exoplr2avp.upstream.StatsDataSource.read(StatsDataSource.java:98)
        at com.google.android.exoplr2avp.extractor.DefaultExtractorInput.readFromUpstream(DefaultExtractorInput.java:299)
        at com.google.android.exoplr2avp.extractor.DefaultExtractorInput.read(DefaultExtractorInput.java:76)
        at com.google.android.exoplr2avp.extractor.ts.TsExtractor.fillBufferWithAtLeastOnePacket(TsExtractor.java:441)
        at com.google.android.exoplr2avp.extractor.ts.TsExtractor.read(TsExtractor.java:331)
        at com.google.android.exoplr2avp.source.hls.BundledHlsMediaChunkExtractor.read(BundledHlsMediaChunkExtractor.java:73)
        at com.google.android.exoplr2avp.source.hls.HlsMediaChunk.feedDataToExtractor(HlsMediaChunk.java:513)
        at com.google.android.exoplr2avp.source.hls.HlsMediaChunk.loadMedia(HlsMediaChunk.java:477)
        at com.google.android.exoplr2avp.source.hls.HlsMediaChunk.load(HlsMediaChunk.java:434)
        at com.google.android.exoplr2avp.upstream.Loader$LoadTask.run(Loader.java:420)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: com.google.android.exoplr2avp.upstream.DataSourceException
        at com.google.android.exoplr2avp.upstream.DefaultHttpDataSource.open(DefaultHttpDataSource.java:417)
        at com.google.android.exoplr2avp.upstream.TeeDataSource.open(TeeDataSource.java:60) 
        at com.google.android.exoplr2avp.upstream.cache.CacheDataSource.openNextSource(CacheDataSource.java:802) 
        at com.google.android.exoplr2avp.upstream.cache.CacheDataSource.read(CacheDataSource.java:657) 
        at com.google.android.exoplr2avp.upstream.StatsDataSource.read(StatsDataSource.java:98) 
        at com.google.android.exoplr2avp.extractor.DefaultExtractorInput.readFromUpstream(DefaultExtractorInput.java:299) 
        at com.google.android.exoplr2avp.extractor.DefaultExtractorInput.read(DefaultExtractorInput.java:76) 
        at com.google.android.exoplr2avp.extractor.ts.TsExtractor.fillBufferWithAtLeastOnePacket(TsExtractor.java:441) 
        at com.google.android.exoplr2avp.extractor.ts.TsExtractor.read(TsExtractor.java:331) 
        at com.google.android.exoplr2avp.source.hls.BundledHlsMediaChunkExtractor.read(BundledHlsMediaChunkExtractor.java:73) 
        at com.google.android.exoplr2avp.source.hls.HlsMediaChunk.feedDataToExtractor(HlsMediaChunk.java:513) 
        at com.google.android.exoplr2avp.source.hls.HlsMediaChunk.loadMedia(HlsMediaChunk.java:477) 
        at com.google.android.exoplr2avp.source.hls.HlsMediaChunk.load(HlsMediaChunk.java:434) 
        at com.google.android.exoplr2avp.upstream.Loader$LoadTask.run(Loader.java:420) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764)

**Comment by Chris-RH** [comment]
Caching is available on Android and iOS platforms.  There are quite specific requirements.  There is more info here: https://renderheads.com/content/docs/AVProVideo/articles/feature-caching.html


"InvalidResponseCodeException: Response code: 416" is a network error, it suggests that the player is requesting an invalid byte range from the server.
What are the specs of your stream?
Are you able to provide a test URL?

**Comment by Winvz** [comment]
I replaced the original test address with mine in the Demo_MediaPlayer scenario,here is my main test code.
I've noticed that the Debug BufferRange only seems to update once a video is playing, even if my log tells me that the video is 100% Cached (CachedMediaStatus = Cached).





Here are two links to my tests.
264:https://v-mps.crazymaplestudios.com/vod-112094/b47ed0607ca271eebf8a5114c1ca0102/6549bac148234afa9eae1c05ffa958e4-41c133e8c4abdb81f61f15bf595d3013-ld.m3u8
265:https://v-mps.crazymaplestudios.com/vod-112094/b47ed0607ca271eebf8a5114c1ca0102/6549bac148234afa9eae1c05ffa958e4-b108a7af2ed97ee12ada9d735d3eed69-ld.m3u8




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "RenderHeads/UnityPlugin-AVProVideo"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年11月16日(星期四) 晚上7:12
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [RenderHeads/UnityPlugin-AVProVideo] About unity version (Issue #1687)





 
Caching is available on Android and iOS platforms.  There are quite specific requirements.  There is more info here: https://renderheads.com/content/docs/AVProVideo/articles/feature-caching.html
 
"InvalidResponseCodeException: Response code: 416" is a network error, it suggests that the player is requesting an invalid byte range from the server.
 What are the specs of your stream?
 Are you able to provide a test URL?

**Comment by Winvz** [comment]
Mail sent

**Comment by Chris-RH** [comment]
I've done a quick android build using Unity 2019.4.40f1 and AVPro Video mediaplayer demo scene and your h264 url.  The only things I changed were in Player Settings:
Graphics API -> OpenGLES3
Internet Access -> Require
It runs absolutely fine.

**Comment by Winvz** [comment]
![image](https://github.com/RenderHeads/UnityPlugin-AVProVideo/assets/48202519/ea3819b3-7ee3-4b7e-8e9b-ef984f702832)

Are your tests cached and then played?
I'll run the tests against your environment

**Comment by Winvz** [comment]
https://github.com/RenderHeads/UnityPlugin-AVProVideo/assets/48202519/8b48a2a5-8d64-49a2-a01d-36e7c77e7f18

This is the latest test video

This happens when I open url1 using the OpenMedia method when the url1 cache progress is less than 100%

**Comment by Chris-RH** [comment]
Ahh, no I was testing that the playback worked, as that is what you were initially asking about.
  
Playback while caching should be possible, but it depends on network resources.
Have you tried playing it from the cache once its been fully downloaded?

**Comment by Winvz** [comment]
In the aforementioned tests, I disabled AutoOpen.

1.Play from the beginning of the video.
2.First, call AddMediaToCache. When the caching progress (p) is greater than 0% and less than 100%, call OpenMedia. The video will play up to the progress of p and then appear to freeze. At this point, the caching progress is generally greater than p, and may even be at 100%. Under these operations, I have noticed that when I start playing the video, the buffer does not refresh anymore.
3.Playback after complete download poses no issues.
